File: df4k-core/src/main/java/org/df4j/core/connector/reactivestream/ReactiveInput.java
package org.df4j.core.connector.reactivestream;

import org.df4j.core.connector.messagestream.StreamInput;
import org.df4j.core.node.AsyncTask;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Iterator;

/**
 * A Queue of tokens of type <T>
 *
 * @param <T>
 */
public class ReactiveInput<T> extends StreamInput<T> implements ReactiveSubscriber<T>, Iterator<T> {
    protected Deque<T> queue;
    protected boolean closeRequested = false;
    protected int capacity;
    protected ReactiveSubscription subscription;

    public ReactiveInput(AsyncTask actor, int capacity) {
        super(actor);
        this.queue = new ArrayDeque<>(capacity);
        this.capacity = capacity;
    }

    public ReactiveInput(AsyncTask actor) {
        this(actor, 8);
    }

    protected int size() {
        return queue.size();
    }

    @Override
    public void onSubscribe(ReactiveSubscription subscription) {
        this.subscription = subscription;
        subscription.request(capacity);
    }

    @Override
    public synchronized void post(T token) {
        if (subscription == null) {
            throw new IllegalStateException("not yet subscribed");
        }
        if (queue.size() >= capacity) {
            throw new IllegalStateException("no space for next token");
        }
        super.post(token);
    }

    @Override
    public synchronized T next() {
        subscription.request(1);
        return super.next();
    }

    public synchronized boolean  isClosed() {
        return closeRequested && (value == null);
    }
}

File: df4k-core/src/main/kotlin/org/df4k/core/connector/reactivestream/ReactiveInput.kt
package org.df4k.core.connector.reactivestream

import org.df4k.core.connector.messagestream.StreamInput
import org.df4k.core.node.AsyncTask

import java.util.ArrayDeque
import java.util.Deque

/**
 * A Queue of tokens of type <T>
 *
 * @param <T>
</T></T> */
class ReactiveInput<T> @JvmOverloads constructor(actor: AsyncTask, protected var capacity: Int = 8) : StreamInput<T>(actor), ReactiveSubscriber<T>, Iterator<T> {
    protected var queue: Deque<T>
    protected var closeRequested = false
    protected var subscription: ReactiveSubscription? = null

    override val isClosed: Boolean
        @Synchronized get() = closeRequested && value == null

    init {
        this.queue = ArrayDeque<T>(capacity)
    }

    override fun size(): Int {
        return queue.size
    }

    override fun onSubscribe(subscription: ReactiveSubscription) {
        this.subscription = subscription
        subscription.request(capacity.toLong())
    }

    @Synchronized
    override fun post(token: T?) {
        if (subscription == null) {
            throw IllegalStateException("not yet subscribed")
        }
        if (queue.size >= capacity) {
            throw IllegalStateException("no space for next token")
        }
        super.post(token)
    }

    @Synchronized
    override fun next(): T? {
        subscription!!.request(1)
        return super.next()
    }
}
